File: r/anagram/anagram.R
library(magrittr)

to_key <- . %>% strsplit("") %>%
                unlist %>%
                sort %>%
                paste0(collapse = "")


anagram <- function(subject, candidates) {
  lc_subj <- tolower(subject)
  key <- to_key(lc_subj)

  # closure: determine if a candidate word is an anagram
  # of the subject
  is_anagram <- function(candidate) {
    lc_cand <- tolower(candidate)
    lc_subj != lc_cand && key == to_key(lc_cand)
  }

  anagrams <- Filter(is_anagram, candidates)
  if (length(anagrams) > 0) anagrams else c()
}

File: julia/anagram/anagram.jl
function detect_anagrams(subject::AbstractString, candidates::AbstractArray)
    tokey = s -> split(s, "") |> sort |> join

    ls = lowercase(subject)
    key = tokey(ls)

    """ take 1
    anagrams = []
    for candidate in candidates
        lc = lowercase(candidate)
        if ls != lc && key == tokey(lc)
            push!(anagrams, candidate)
        end
    end
    anagrams
    """

    # take 2
    filter(candidates) do candidate
        lc = lowercase(candidate)
        ls != lc && key == tokey(lc)
    end
end
